Windows builds on CI are failing with EmitC enabled

Windows builds on our GitHub Actions CI are failing after enabling EmitC by default (https://github.com/google/iree/pull/6717): https://github.com/google/iree/runs/3299480891?check_suite_focus=true

Logs are a bunch of this:
```
FAILED: iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_module_test.dir/module_test.cc.obj 
  C:\PROGRA~2\MICROS~1\2019\ENTERP~1\VC\Tools\MSVC\1429~1.300\bin\Hostx64\x64\cl.exe  /nologo /TP -DIREE_HAVE_EMITC_DIALECT -ID:\a\iree\iree\main_checkout -ID:\a\iree\iree\iree-build -ID:\a\iree\iree\main_checkout\third_party\googletest\googlemock\include -ID:\a\iree\iree\main_checkout\third_party\googletest\googlemock -ID:\a\iree\iree\main_checkout\third_party\googletest\googletest\include -ID:\a\iree\iree\main_checkout\third_party\googletest\googletest /DWIN32 /D_WINDOWS  /EHsc /MD /O2 /Ob2 /DNDEBUG /W3 /DWIN32_LEAN_AND_MEAN /DNOMINMAX /D_USE_MATH_DEFINES /D_CRT_SECURE_NO_WARNINGS /D_CRT_SECURE_CPP_OVERLOAD_STANDARD_NAMES /GR- /bigobj /wd4200 /wd4018 /wd4146 /wd4244 /wd4267 /wd4005 /wd4065 /wd4141 /wd4624 /wd5105 -std:c++14 /showIncludes /Foiree\vm\test\emitc\CMakeFiles\iree_vm_test_emitc_module_test.dir\module_test.cc.obj /Fdiree\vm\test\emitc\CMakeFiles\iree_vm_test_emitc_module_test.dir\ /FS -c D:\a\iree\iree\main_checkout\iree\vm\test\emitc\module_test.cc
  D:\a\iree\iree\iree-build\iree/vm/test/emitc/arithmetic_ops.h(16): error C2229: struct 'arithmetic_ops_state_t' has an illegal zero-sized array
  D:\a\iree\iree\iree-build\iree/vm/test/emitc/arithmetic_ops.h(17): error C2229: struct 'arithmetic_ops_state_t' has an illegal zero-sized array
  D:\a\iree\iree\iree-build\iree/vm/test/emitc/arithmetic_ops.h(18): error C2229: struct 'arithmetic_ops_state_t' has an illegal zero-sized array
  D:\a\iree\iree\iree-build\iree/vm/test/emitc/arithmetic_ops.h(736): error C2466: cannot allocate an array of constant size 0
  D:\a\iree\iree\iree-build\iree/vm/test/emitc/arithmetic_ops_f32.h(16): error C2229: struct 'arithmetic_ops_f32_state_t' has an illegal zero-sized array
  D:\a\iree\iree\iree-build\iree/vm/test/emitc/arithmetic_ops_f32.h(17): error C2229: struct 'arithmetic_ops_f32_state_t' has an illegal zero-sized array
  D:\a\iree\iree\iree-build\iree/vm/test/emitc/arithmetic_ops_f32.h(18): error C2229: struct 'arithmetic_ops_f32_state_t' has an illegal zero-sized array
  D:\a\iree\iree\iree-build\iree/vm/test/emitc/arithmetic_ops_f32.h(1473): error C2466: cannot allocate an array of constant size 0
```

cc @marbre

--- a/iree/compiler/Dialect/VM/Conversion/VMToEmitC/DropExcludedExports.cpp
+++ b/iree/compiler/Dialect/VM/Conversion/VMToEmitC/DropExcludedExports.cpp
@@ -31,7 +31,6 @@ class DropExcludedExportsPass
     getOperation()->walk([&](IREE::VM::ExportOp exportOp) {
       Operation *op = exportOp.getOperation();
       if (op->hasAttr("emitc.exclude")) {
-        exportOp.emitRemark("Dropping excluded export");
         opsToRemove.push_back(op);
       }
     });

--- a/iree/compiler/Dialect/VM/Target/C/CModuleTarget.cpp
+++ b/iree/compiler/Dialect/VM/Target/C/CModuleTarget.cpp
@@ -23,6 +23,16 @@ namespace iree_compiler {
 namespace IREE {
 namespace VM {
 
+static void printCompilerConfigurationBlock(llvm::raw_ostream &output) {
+  output << "//" << std::string(77, '=') << "\n"
+         << "// compiler configuration\n"
+         << "//" << std::string(77, '=') << "\n\n";
+
+  output << "#if defined(IREE_COMPILER_MSVC)\n";
+  output << "#pragma warning(disable:4102)\n";
+  output << "#endif  // IREE_COMPILER_MSVC\n";
+}
+
 static void printModuleComment(IREE::VM::ModuleOp &moduleOp,
                                llvm::raw_ostream &output) {
   output << "//" << std::string(77, '=') << "\n"
@@ -41,6 +51,7 @@ static void printSeparatingComment(llvm::raw_ostream &output) {
             "//"
          << std::string(77, '=') << "\n";
 }
+
 static LogicalResult printRodataBuffers(IREE::VM::ModuleOp &moduleOp,
                                         mlir::emitc::CppEmitter &emitter) {
   llvm::raw_ostream &output = emitter.ostream();
@@ -81,6 +92,7 @@ static LogicalResult printRodataBuffers(IREE::VM::ModuleOp &moduleOp,
 
   return success();
 }
+
 static LogicalResult printStructDefinitions(IREE::VM::ModuleOp &moduleOp,
                                             mlir::emitc::CppEmitter &emitter) {
   llvm::raw_ostream &output = emitter.ostream();
@@ -89,15 +101,21 @@ static LogicalResult printStructDefinitions(IREE::VM::ModuleOp &moduleOp,
   output << "struct " << moduleName << "_t {};\n";
   output << "struct " << moduleName << "_state_t {\n";
 
+  // Returns |count| or 1 if |count| == 0.
+  // Some compilers (MSVC) don't support zero-length struct fields on the
+  // interior of structs (just VLA at the tail).
+  auto countOrEmpty = [](uint32_t count) { return count ? count : 1; };
+
+  auto ordinalCounts = moduleOp.ordinal_counts().getValue();
   output << "iree_allocator_t allocator;\n";
-  output << "uint8_t rwdata["
-         << moduleOp.ordinal_counts().getValue().global_bytes() << "];\n";
-  output << "iree_vm_ref_t refs["
-         << moduleOp.ordinal_counts().getValue().global_refs() << "];\n";
+  output << "uint8_t rwdata[" << countOrEmpty(ordinalCounts.global_bytes())
+         << "];\n";
+  output << "iree_vm_ref_t refs[" << countOrEmpty(ordinalCounts.global_refs())
+         << "];\n";
   output << "iree_vm_buffer_t rodata_buffers["
-         << moduleOp.ordinal_counts().getValue().rodatas() << "];\n";
+         << countOrEmpty(ordinalCounts.rodatas()) << "];\n";
   output << "iree_vm_function_t imports["
-         << moduleOp.ordinal_counts().getValue().import_funcs() << "];\n";
+         << countOrEmpty(ordinalCounts.import_funcs()) << "];\n";
   output << "};\n";
 
   output << "typedef struct " << moduleName << "_t " << moduleName << "_t;\n";
@@ -150,76 +168,95 @@ static LogicalResult buildModuleDescriptors(IREE::VM::ModuleOp &moduleOp,
   };
 
   // exports
-  std::string exportName = moduleName + "_exports_";
-  output << "static const iree_vm_native_export_descriptor_t " << exportName
-         << "[] = {\n";
-
-  // sort export ops
   SmallVector<IREE::VM::ExportOp, 4> exportOps(
       moduleOp.getOps<IREE::VM::ExportOp>());
-  llvm::sort(exportOps, [](auto &lhs, auto &rhs) {
-    return lhs.export_name().compare(rhs.export_name()) < 0;
-  });
+  std::string exportName = moduleName + "_exports_";
+  output << "static const size_t " << exportName
+         << "count_ = " << exportOps.size() << ";\n";
+  output << "static const iree_vm_native_export_descriptor_t " << exportName
+         << "[] = {\n";
+  if (exportOps.empty()) {
+    // Empty list placeholder.
+    output << "    {0},\n";
+  } else {
+    // sort export ops
+    llvm::sort(exportOps, [](auto &lhs, auto &rhs) {
+      return lhs.export_name().compare(rhs.export_name()) < 0;
+    });
+    for (auto exportOp : exportOps) {
+      StringRef funcName = exportOp.function_ref();
+      auto funcOp = symbolTable.lookup<mlir::FuncOp>(funcName);
+      if (!funcOp) {
+        return exportOp.emitError("Couldn't find referenced FuncOp");
+      }
+      StringAttr callingConvention = funcOp.getOperation()
+                                         ->getAttr("calling_convention")
+                                         .cast<StringAttr>();
+      if (!callingConvention) {
+        return exportOp.emitError(
+            "Couldn't create calling convention string for referenced FuncOp");
+      }
 
-  for (auto exportOp : exportOps) {
-    StringRef funcName = exportOp.function_ref();
-    auto funcOp = symbolTable.lookup<mlir::FuncOp>(funcName);
-    if (!funcOp) {
-      return exportOp.emitError("Couldn't find referenced FuncOp");
+      // TODO(simon-camp): support function-level reflection attributes
+      output << "{" << printCStringView(exportOp.export_name()) << ", "
+             << printCStringView(callingConvention.getValue())
+             << ", 0, NULL},\n";
     }
-    StringAttr callingConvention =
-        funcOp.getOperation()->getAttr("calling_convention").cast<StringAttr>();
-    if (!callingConvention) {
-      return exportOp.emitError(
-          "Couldn't create calling convention string for referenced FuncOp");
-    }
-
-    // TODO(simon-camp): support function-level reflection attributes
-    output << "{" << printCStringView(exportOp.export_name()) << ", "
-           << printCStringView(callingConvention.getValue()) << ", 0, NULL},\n";
   }
   output << "};\n";
   output << "\n";
 
   // imports
-  std::string importName = moduleName + "_imports_";
-  output << "static const iree_vm_native_import_descriptor_t " << importName
-         << "[] = {\n";
-
-  // sort import ops
   SmallVector<IREE::VM::ImportOp, 4> importOps(
       moduleOp.getOps<IREE::VM::ImportOp>());
-  llvm::sort(importOps, [](auto &lhs, auto &rhs) {
-    return lhs.getName().compare(rhs.getName()) < 0;
-  });
-
-  for (auto importOp : importOps) {
-    output << "{" << printCStringView(importOp.getName()) << "},\n";
+  std::string importName = moduleName + "_imports_";
+  output << "static const size_t " << importName
+         << "count_ = " << importOps.size() << ";\n";
+  output << "static const iree_vm_native_import_descriptor_t " << importName
+         << "[] = {\n";
+  if (importOps.empty()) {
+    // Empty list placeholder.
+    output << "    {0},\n";
+  } else {
+    // sort import ops
+    llvm::sort(importOps, [](auto &lhs, auto &rhs) {
+      return lhs.getName().compare(rhs.getName()) < 0;
+    });
+    for (auto importOp : importOps) {
+      output << "{" << printCStringView(importOp.getName()) << "},\n";
+    }
   }
   output << "};\n";
   output << "\n";
 
   // functions
   std::string functionName = moduleName + "_funcs_";
+  output << "static const size_t " << functionName
+         << "count_ = " << exportOps.size() << ";\n";
   output << "static const iree_vm_native_function_ptr_t " << functionName
          << "[] = {\n";
+  if (exportOps.empty()) {
+    // Empty list placeholder.
+    output << "    {0},\n";
+  } else {
+    // We only add exported functions to the table, as calls to internal
+    // functions are directly mapped to C function calls of the generated
+    // implementation.
+    for (auto exportOp : exportOps) {
+      StringRef funcName = exportOp.function_ref();
+      auto funcOp = symbolTable.lookup<mlir::FuncOp>(funcName);
+      if (!funcOp) {
+        return exportOp.emitError("Couldn't find referenced FuncOp");
+      }
+      output << "{"
+             << "(iree_vm_native_function_shim_t)";
 
-  // We only add exported functions to the table, as calls to internal functions
-  // are directly mapped to C function calls of the generated implementation.
-  for (auto exportOp : exportOps) {
-    StringRef funcName = exportOp.function_ref();
-    auto funcOp = symbolTable.lookup<mlir::FuncOp>(funcName);
-    if (!funcOp) {
-      return exportOp.emitError("Couldn't find referenced FuncOp");
+      if (failed(printShim(funcOp, output))) {
+        return funcOp.emitError("Couldn't create calling convention string");
+      }
+      output << ", "
+             << "(iree_vm_native_function_target_t)" << funcName << "},\n";
     }
-    output << "{"
-           << "(iree_vm_native_function_shim_t)";
-
-    if (failed(printShim(funcOp, output))) {
-      return funcOp.emitError("Couldn't create calling convention string");
-    }
-    output << ", "
-           << "(iree_vm_native_function_target_t)" << funcName << "},\n";
   }
   output << "};\n";
   output << "\n";
@@ -230,11 +267,11 @@ static LogicalResult buildModuleDescriptors(IREE::VM::ModuleOp &moduleOp,
   output << "static const iree_vm_native_module_descriptor_t " << descriptorName
          << " = {\n"
          << printCStringView(moduleName) << ",\n"
-         << "IREE_ARRAYSIZE(" << importName << "),\n"
+         << importName << "count_,\n"
          << importName << ",\n"
-         << "IREE_ARRAYSIZE(" << exportName << "),\n"
+         << exportName << "count_,\n"
          << exportName << ",\n"
-         << "IREE_ARRAYSIZE(" << functionName << "),\n"
+         << functionName << "count_,\n"
          << functionName << ",\n"
          << "0,\n"
          << "NULL,\n"
@@ -404,6 +441,9 @@ LogicalResult translateModuleToC(IREE::VM::ModuleOp moduleOp,
   printInclude("iree/vm/ops_emitc.h");
   printInclude("iree/vm/shims_emitc.h");
   printInclude("iree/vm/value.h");
+  output << "\n";
+
+  printCompilerConfigurationBlock(output);
   output << "\n";
 
   printModuleComment(moduleOp, output);

--- a/iree/compiler/Dialect/VM/Target/C/test/constant_ops.mlir
+++ b/iree/compiler/Dialect/VM/Target/C/test/constant_ops.mlir
@@ -9,10 +9,10 @@ vm.module @constant_ops {
   // Check the generated state struct
   // CHECK-LABEL: struct constant_ops_state_t {
   // CHECK-NEXT: iree_allocator_t allocator;
-  // CHECK-NEXT: uint8_t rwdata[0];
-  // CHECK-NEXT: iree_vm_ref_t refs[0];
+  // CHECK-NEXT: uint8_t rwdata[1];
+  // CHECK-NEXT: iree_vm_ref_t refs[1];
   // CHECK-NEXT: iree_vm_buffer_t rodata_buffers[2];
-  // CHECK-NEXT: iree_vm_function_t imports[0];
+  // CHECK-NEXT: iree_vm_function_t imports[1];
   // CHECK-NEXT: };
 
   // We mark the rodata ops public in this test to explicitly prevent DCE from

--- a/iree/compiler/Dialect/VM/Target/C/test/global_ops.mlir
+++ b/iree/compiler/Dialect/VM/Target/C/test/global_ops.mlir
@@ -5,9 +5,9 @@ vm.module @global_ops {
   // CHECK-LABEL: struct global_ops_state_t {
   // CHECK-NEXT: iree_allocator_t allocator;
   // CHECK-NEXT: uint8_t rwdata[8];
-  // CHECK-NEXT: iree_vm_ref_t refs[0];
-  // CHECK-NEXT: iree_vm_buffer_t rodata_buffers[0];
-  // CHECK-NEXT: iree_vm_function_t imports[0];
+  // CHECK-NEXT: iree_vm_ref_t refs[1];
+  // CHECK-NEXT: iree_vm_buffer_t rodata_buffers[1];
+  // CHECK-NEXT: iree_vm_function_t imports[1];
   // CHECK-NEXT: };
 
   vm.global.i32 mutable @c42 = 42 : i32

